### PR TITLE
Add optional cloudsql auth proxy sidecar for applications

### PR DESCRIPTION
**Pull request by martinhny**

### CODE REVIEW

File: tests/application/cloudsql-auth-proxy/set-version.yaml
Comment (evenh) [inline tests/application/cloudsql-auth-proxy/set-version.yaml:1]: This file and the assert is not included in `chainsaw-test.yaml`

File: tests/application/files-from/cloudsql-auth-proxy/chainsaw-test.yaml
Comment (evenh) [inline tests/application/files-from/cloudsql-auth-proxy/chainsaw-test.yaml:1]: What's the use case for this folder and files?

**Review by evenh — CHANGES_REQUESTED** [review]
Missing support for CloudSQL for jobs. I think that would be a big use case for batch jobs etc.

File: api/v1alpha1/application_types.go [hunk 1]
```diff
@@ -246,6 +246,30 @@ type ApplicationSpec struct {
...
+}
+
+type CloudSQLSettings struct {
+	// Enables the sidecar for CloudSQL proxy
+
+	//+kubebuilder:validation:Optional
+	//+kubebuilder:default:=false
+	Enabled bool `json:"enabled,omitempty"`
+
+	// Connection name for the CloudSQL instance. Found in the Google Cloud Console.
```
[the remark this answers is not in the document]
Comment (martinhny) [reply api/v1alpha1/application_types.go]: added

File: pkg/resourcegenerator/core/pod.go [hunk 2]
```diff
@@ -93,6 +94,36 @@ func CreateApplicationContainer(application *skiperatorv1alpha1.Application, opt
...
+			RunAsGroup:               util.PointToInt64(200),
+			ReadOnlyRootFilesystem:   util.PointTo(true),
+			AllowPrivilegeEscalation: util.PointTo(false),
+			SeccompProfile: &corev1.SeccompProfile{
+				Type: "RuntimeDefault",
+			},
+		},
+		Resources: corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				corev1.ResourceMemory: resource.MustParse("64Mi"),
```
[the remark this answers is not in the document]
Comment (martinhny) [reply pkg/resourcegenerator/core/pod.go:120]: We don't know enough about it yet

File: api/v1alpha1/application_types.go [hunk 3]
```diff
@@ -246,6 +246,30 @@ type ApplicationSpec struct {
...
+
+	//+kubebuilder:validation:Required
```
[the remark this answers is not in the document]
Comment (martinhny) [reply api/v1alpha1/application_types.go:266]: added

File: tests/application/cloudsql-auth-proxy/set-version.yaml
Comment (martinhny) [reply tests/application/cloudsql-auth-proxy/set-version.yaml:1]: fixd

File: tests/application/files-from/cloudsql-auth-proxy/chainsaw-test.yaml
Comment (martinhny) [reply tests/application/files-from/cloudsql-auth-proxy/chainsaw-test.yaml:1]: mistakenly copy pasted it :) Removed!

**Comment by martinhny** [comment]
We will wait until native support for sidecars before we support cloudsql on skipjobs. https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/
Draft: https://github.com/kartverket/skiperator/pull/457

**Comment by eliihen** [comment]
Suggestion: Look at what naiserqtor does in terms of API

https://docs.nais.io/reference/application-spec/#gcpsqlinstances

**Comment by martinhny** [comment]
@eliihen Do you think it would be more suitable to put this configuration under gcp? 
ie:
```
gcp:
  cloudSql: 
```
Other than that it looks like naiserator can create cloud resources, which is outside of the scope of this pull request.

**Comment by eliihen** [comment]
Good idea! How about putting this under a `gcp` section that makes it clearer that this just creates a sidecar proxy? I kind of like the approach that nais has here to abstract over the creation of cloud resources, so maybe that's a feature we could look into in the future. 

```
gcp:
  cloudSqlProxy: 
    ...
```

File: pkg/resourcegenerator/istio/service_entry.go [hunk 4]
```diff
@@ -14,9 +14,43 @@ import (
...
+	if !util.IsCloudSqlProxyEnabled(application.Spec.GCP) {
+		return accessPolicy, nil
+	}
+
+	if application.Spec.GCP.CloudSQLProxy.IP == "" {
+		return nil, errors.New("cloud sql proxy IP is not set")
+	}
+
+	externalRule := &podtypes.ExternalRule{
+		Host:  fmt.Sprintf("%x.cloudsql", util.GenerateHashFromName(application.GetName())),
```
Comment (evenh) [inline pkg/resourcegenerator/istio/service_entry.go:32]: I would love to have something more descriptive, just for making debugging Istio easier. I understand that this field is set in order to make Istio validation pass, but this hostname might appear in the logs. What about `appname-${hash(appName-image:tag}` in order to make them more distinct across various deployed versions of the app (e.g. preview environments)?

**Review by evenh — APPROVED** [review]

File: pkg/resourcegenerator/istio/service_entry.go (hunk 4 above)
Comment (martinhny) [reply pkg/resourcegenerator/istio/service_entry.go]: I changed the name to `appName-image.cloudsql` so its more descriptive.

Comment (martinhny) [reply pkg/resourcegenerator/istio/service_entry.go]: done :+1:

**Pull request merged**